bookmarks/tests/test_context_path.py

```diff
@@ -0,0 +1,42 @@
+import datetime
+import email
+import importlib
+import os
+import urllib.parse
+
+from django.test import TestCase, override_settings
+from django.urls import reverse, include, resolvers
+from django.urls.resolvers import URLResolver, RegexPattern
+from django.conf import settings
+
+
+class ContextPathTestCase(TestCase):
+
+    def setUp(self):
+        self.bookmarks_urls = importlib.import_module('bookmarks.urls')
+        self.siteroot_urls = importlib.import_module('siteroot.urls')
+
+    @override_settings(LD_CONTEXT_PATH=None)
+    def tearDown(self):
+        importlib.reload(self.bookmarks_urls)
+        importlib.reload(self.siteroot_urls)
+
+    @override_settings(LD_CONTEXT_PATH='linkding/')
+    def test_route_with_context_path(self):
+        bookmarks_patterns = importlib.reload(self.bookmarks_urls).urlpatterns
+        bookmarks_match = bookmarks_patterns[0].resolve('linkding/bookmarks')
+        self.assertEqual(bookmarks_match.url_name, 'index')
```

Comment:
Would be great if we could test that the `django.urls.reverse` function returns the correct results, as this is the method that is practically used everywhere in the codebase, and I guess also in the `url` template tag.

So for example
```python
from django.urls import reverse

...
    # in the test
    url = reverse('bookmarks:index')
    self.assertEqual(url, 'linkding/bookmarks')
```

I could not get it to work yet, but maybe there is a way. Not a blocker though.

Reply:
So you can pass a specific `urlconf` to `reverse`, which allows using dynamic URL configs. The `reverse` function actually caches the url config, so passing the actual Python module does not work for both test cases, as it will be cached, even with the `importlib.reload` call. This can be solved by created a mock url config that just contains the `urlpatterns` attribute that Django uses for resolving URLs.

In summary this is what I came up with for the tests:
```python
import importlib

from django.test import TestCase, override_settings
from django.urls import reverse


class MockUrlConf:
    def __init__(self, module):
        self.urlpatterns = module.urlpatterns


class ContextPathTestCase(TestCase):

    def setUp(self):
        self.siteroot_urls = importlib.import_module('siteroot.urls')

    @override_settings(LD_CONTEXT_PATH=None)
    def tearDown(self):
        importlib.reload(self.siteroot_urls)

    @override_settings(LD_CONTEXT_PATH='linkding/')
    def test_route_with_context_path(self):
        module = importlib.reload(self.siteroot_urls)
        # pass mock config instead of actual module to prevent caching the
        # url config in django.urls.reverse
        urlconf = MockUrlConf(module)
        test_cases = [
            ('bookmarks:index', '/linkding/bookmarks'),
            ('bookmarks:bookmark-list', '/linkding/api/bookmarks/'),
            ('login', '/linkding/login/'),
            ('admin:bookmarks_bookmark_changelist', '/linkding/admin/bookmarks/bookmark/'),
        ]

        for url_name, expected_url in test_cases:
            url = reverse(url_name, urlconf=urlconf)
            self.assertEqual(expected_url, url)

    @override_settings(LD_CONTEXT_PATH='')
    def test_route_without_context_path(self):
        module = importlib.reload(self.siteroot_urls)
        # pass mock config instead of actual module to prevent caching the
        # url config in django.urls.reverse
        urlconf = MockUrlConf(module)
        test_cases = [
            ('bookmarks:index', '/bookmarks'),
            ('bookmarks:bookmark-list', '/api/bookmarks/'),
            ('login', '/login/'),
            ('admin:bookmarks_bookmark_changelist', '/admin/bookmarks/bookmark/'),
        ]

        for url_name, expected_url in test_cases:
            url = reverse(url_name, urlconf=urlconf)
            self.assertEqual(expected_url, url)
```

Reply:
This is exactly what I want, but I was unable to bypass django's cache before.
I will use your test case instead.